modules/edu.gemini.seqexec.web/edu.gemini.seqexec.web.shared/src/main/scala/edu/gemini/seqexec/web/common/FixedLengthBuffer.scala

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2016-2017 Association of Universities for Research in Astronomy, Inc. (AURA)
+// For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+package edu.gemini.web.common
+
+import scalaz.{Show, Equal, Functor}
+import scalaz.syntax.std.boolean._
+
+object FixedLengthBuffer {
+  private final case class FixedLengthBufferImpl[A](maxLength: Int, data: Vector[A]) extends FixedLengthBuffer[A] {
```

Comment:
I've seen this pattern `...Impl` a lot in Scala, but I'm still not sure what's for? Is it to hide the default constructors and create your own custom *smart* constructors?

Reply:
exactly. since the `Impl` class is private you cannot see so you have to go through the `apply` or `unsafe...` constructors. For this particular class it means we can restrict the `maxLength` to only positive values

Now that you mention it, there is a slightly cleaner technique with `abstract class`. i'll try that before merging